In [1]:
import io
import pandas as pd
import requests

url = "https://mydataaijournal.com/public/andy-data/olist-data/orders_milestone.parquet"

print("Fetching dataset from mydataaijournal.com...")
response = requests.get(url)
response.raise_for_status()

df_milestones = pd.read_parquet(io.BytesIO(response.content))
print(f"Successfully loaded {len(df_milestones):,} rows into memory!")

Fetching dataset from mydataaijournal.com...
Successfully loaded 392,856 rows into memory!


In [2]:
df_milestones.head()

,ORDER_ID,CUSTOMER_ID,MILESTONE_NAME,MILESTONE_TIMESTAMP,PREVIOUS_MILESTONE_TIMESTAMP,HOURS_SINCE_LAST_STEP
0,00010242fe8c5a6d1ba2dd792cb16214,3ce436f183e68e07877b285a838db11a,1_purchased,2017-09-13 08:59:02,NaT,NaN
1,00010242fe8c5a6d1ba2dd792cb16214,3ce436f183e68e07877b285a838db11a,2_approved,2017-09-13 09:45:35,2017-09-13 08:59:02,0.775833
2,00010242fe8c5a6d1ba2dd792cb16214,3ce436f183e68e07877b285a838db11a,3_carrier_pickup,2017-09-19 18:34:16,2017-09-13 09:45:35,152.811389
3,00010242fe8c5a6d1ba2dd792cb16214,3ce436f183e68e07877b285a838db11a,4_delivered,2017-09-20 23:43:48,2017-09-19 18:34:16,29.158889
4,00018f77f2f0320c557190d7a144bdd3,f6dd3ec061db4e3987629fe6b26e5cce,1_purchased,2017-04-26 10:53:06,NaT,NaN


In [3]:
df_milestones.columns = df_milestones.columns.str.lower()

df_milestones.head()

,order_id,customer_id,milestone_name,milestone_timestamp,previous_milestone_timestamp,hours_since_last_step
0,00010242fe8c5a6d1ba2dd792cb16214,3ce436f183e68e07877b285a838db11a,1_purchased,2017-09-13 08:59:02,NaT,NaN
1,00010242fe8c5a6d1ba2dd792cb16214,3ce436f183e68e07877b285a838db11a,2_approved,2017-09-13 09:45:35,2017-09-13 08:59:02,0.775833
2,00010242fe8c5a6d1ba2dd792cb16214,3ce436f183e68e07877b285a838db11a,3_carrier_pickup,2017-09-19 18:34:16,2017-09-13 09:45:35,152.811389
3,00010242fe8c5a6d1ba2dd792cb16214,3ce436f183e68e07877b285a838db11a,4_delivered,2017-09-20 23:43:48,2017-09-19 18:34:16,29.158889
4,00018f77f2f0320c557190d7a144bdd3,f6dd3ec061db4e3987629fe6b26e5cce,1_purchased,2017-04-26 10:53:06,NaT,NaN


In [4]:
df_milestones = df_milestones[df_milestones["milestone_name"]!="1_purchased"]

df_milestones.head()

,order_id,customer_id,milestone_name,milestone_timestamp,previous_milestone_timestamp,hours_since_last_step
1,00010242fe8c5a6d1ba2dd792cb16214,3ce436f183e68e07877b285a838db11a,2_approved,2017-09-13 09:45:35,2017-09-13 08:59:02,0.775833
2,00010242fe8c5a6d1ba2dd792cb16214,3ce436f183e68e07877b285a838db11a,3_carrier_pickup,2017-09-19 18:34:16,2017-09-13 09:45:35,152.811389
3,00010242fe8c5a6d1ba2dd792cb16214,3ce436f183e68e07877b285a838db11a,4_delivered,2017-09-20 23:43:48,2017-09-19 18:34:16,29.158889
5,00018f77f2f0320c557190d7a144bdd3,f6dd3ec061db4e3987629fe6b26e5cce,2_approved,2017-04-26 11:05:13,2017-04-26 10:53:06,0.201944
6,00018f77f2f0320c557190d7a144bdd3,f6dd3ec061db4e3987629fe6b26e5cce,3_carrier_pickup,2017-05-04 14:35:00,2017-04-26 11:05:13,195.496389


In [5]:
df_milestones.info()

<class 'pandas.core.frame.DataFrame'>
Index: 293415 entries, 1 to 392855
Data columns (total 6 columns):
 #   Column                        Non-Null Count   Dtype         
---  ------                        --------------   -----         
 0   order_id                      293415 non-null  object        
 1   customer_id                   293415 non-null  object        
 2   milestone_name                293415 non-null  object        
 3   milestone_timestamp           293415 non-null  datetime64[ns]
 4   previous_milestone_timestamp  293249 non-null  datetime64[ns]
 5   hours_since_last_step         293249 non-null  float64       
dtypes: datetime64[ns](2), float64(1), object(3)
memory usage: 15.7+ MB


In [6]:
df_milestones = (
    df_milestones.dropna(subset=["hours_since_last_step"])
)

df_milestones.info()

<class 'pandas.core.frame.DataFrame'>
Index: 293249 entries, 1 to 392855
Data columns (total 6 columns):
 #   Column                        Non-Null Count   Dtype         
---  ------                        --------------   -----         
 0   order_id                      293249 non-null  object        
 1   customer_id                   293249 non-null  object        
 2   milestone_name                293249 non-null  object        
 3   milestone_timestamp           293249 non-null  datetime64[ns]
 4   previous_milestone_timestamp  293249 non-null  datetime64[ns]
 5   hours_since_last_step         293249 non-null  float64       
dtypes: datetime64[ns](2), float64(1), object(3)
memory usage: 15.7+ MB


In [7]:



distribution_summary = (
    df_milestones.groupby("milestone_name")["hours_since_last_step"]
                 .describe(percentiles=[0.5,0.75,0.95,0.99])
)

distribution_summary

,count,mean,std,min,50%,75%,95%,99%,max
milestone_name,,,,,,,,,
2_approved,99281.0,9.940303,25.084564,0.000000,0.342778,14.064722,46.890556,86.244833,4509.180556
3_carrier_pickup,97492.0,68.210448,83.521482,0.004167,44.218333,86.342292,194.284083,411.378669,3018.301667
4_delivered,96476.0,223.706648,209.878953,0.000000,170.274028,288.651181,580.786806,983.591319,4924.583333


<div style="width:80%; margin:auto; font-size: 1.1rem;">

# 📊 Logistics: Delivery Bottleneck Analysis

> **Executive Summary:** The table breaks down the fulfillment pipeline to show exactly where delays happen. Instead of tracking the total time since checkout, it measures the distinct time spent between each step of the process.
<br><br>
* **Rows (The Stages):** Each row represents a specific, isolated phase of the fulfillment journey (e.g., payment approval wait time, warehouse packing time before carrier handoff, and actual transit time). <br> <br>
* **Columns (Percentile Time Spent):** The percentiles of hours elapsed since the previous step. They show the distribution of phase durations across all orders—ranging from typical duration (50%) to worst-case bottlenecks (99%).
---

### ⏱️ The Fulfillment Pipeline Data

| Milestone | 50% (Median) | 75% | 95% | 99% |
| :--- | :--- | :--- | :--- | :--- |
| ![Approved](https://img.shields.io/badge/-2__approved-success?style=flat-square) | **0.34 hrs** *(~20 mins)* | 14.06 hrs | 46.89 hrs *(~1.95 days)* | 86.24 hrs *(~3.59 days)* |
| ![Carrier Pickup](https://img.shields.io/badge/-carrier__pickup-warning?style=flat-square) | **44.22 hrs** *(~1.84 days)* | 86.34 hrs | 194.28 hrs *(~8.10 days)* | 411.38 hrs *(~17.14 days)* |
| ![Delivered](https://img.shields.io/badge/-4__delivered-critical?style=flat-square) | **170.27 hrs** *(~7.10 days)*| 288.65 hrs | 580.79 hrs *(~24.20 days)* | 983.59 hrs *(~40.98 days)* |

</div>